
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Experiment Quality Assessment &#8212; runHiC 0.8.4 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Release Notes" href="changelog.html" />
    <link rel="prev" title="Parallel Computation" href="parallel.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">runHiC</a></h1>



<p class="blurb">A user-friendly Hi-C data processing software supporting distributed computation</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=XiaoTaoWang&repo=HiC_pipeline&type=star&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Install runHiC</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel.html">Parallel Computation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Experiment Quality Assessment</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#statistic-table">Statistic Table</a></li>
<li class="toctree-l2"><a class="reference internal" href="#library-size-estimation">Library-size Estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#read-pair-type-plotting">Read-pair Type Plotting</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Release Notes</a></li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="experiment-quality-assessment">
<h1>Experiment Quality Assessment<a class="headerlink" href="#experiment-quality-assessment" title="Permalink to this headline">¶</a></h1>
<p>In this tutorial, I will show you how runHiC can be used in data quality assessment.</p>
<p>All you need to type in is a sinlge-line command after running <em>runHiC filtering</em> or <em>runHiC pileup</em>
(refer to <a class="reference external" href="http://xiaotaowang.github.io/HiC_pipeline/quickstart.html">quickstart</a> for more details):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC quality -m datasets.tsv -L filtered-hg38
</pre></div>
</div>
<section id="statistic-table">
<h2>Statistic Table<a class="headerlink" href="#statistic-table" title="Permalink to this headline">¶</a></h2>
<p>Statistic tables on sequencing reads can be found in <em>.stats files under *filtered-hg38</em></p>
<p>Here’s a snapshot:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">000</span><span class="n">_SequencedReads</span><span class="p">:</span>   <span class="mi">14332993</span>
        <span class="mi">010</span><span class="n">_DoubleSideMappedReads</span><span class="p">:</span>  <span class="mi">10323445</span>
        <span class="mi">020</span><span class="n">_SingleSideMappedReads</span><span class="p">:</span>  <span class="mi">3238504</span>
        <span class="mi">030</span><span class="n">_UnmappedReads</span><span class="p">:</span>  <span class="mi">771044</span>
<span class="mi">100</span><span class="n">_NormalPairs</span><span class="p">:</span>  <span class="mi">10323445</span>
        <span class="mi">110</span><span class="n">_AfterFilteringReads</span><span class="p">:</span>  <span class="mi">9050818</span>
        <span class="mi">120</span><span class="n">_SameFragmentReads</span><span class="p">:</span>  <span class="mi">1193185</span>
                <span class="mi">122</span><span class="n">_SelfLigationReads</span><span class="p">:</span>  <span class="mi">245343</span>
                <span class="mi">124</span><span class="n">_DanglingReads</span><span class="p">:</span>  <span class="mi">934456</span>
                <span class="mi">126</span><span class="n">_UnknownMechanism</span><span class="p">:</span>  <span class="mi">13386</span>
        <span class="mi">130</span><span class="n">_DuplicateRemoved</span><span class="p">:</span>  <span class="mi">79442</span>
<span class="mi">400</span><span class="n">_TotalContacts</span><span class="p">:</span>  <span class="mi">9050818</span>
        <span class="mi">410</span><span class="n">_IntraChromosomalReads</span><span class="p">:</span>  <span class="mi">4170313</span>
                <span class="mi">412</span><span class="n">_IntraLongRangeReads</span><span class="p">(</span><span class="o">&gt;=</span><span class="mi">20</span><span class="n">Kb</span><span class="p">):</span>  <span class="mi">2812664</span>
                <span class="mi">412</span><span class="n">_IntraShortRangeReads</span><span class="p">(</span><span class="o">&lt;</span><span class="mi">20</span><span class="n">Kb</span><span class="p">):</span>  <span class="mi">1357649</span>
        <span class="mi">420</span><span class="n">_InterChromosomalReads</span><span class="p">:</span>  <span class="mi">4880505</span>

<span class="n">Critical</span> <span class="n">Indicators</span><span class="p">:</span>
<span class="n">Double</span> <span class="n">Unique</span> <span class="n">Mapped</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">10323445</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.7203</span>
<span class="n">Self</span><span class="o">-</span><span class="n">Ligation</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">245343</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.0171</span>
<span class="n">Dangling</span><span class="o">-</span><span class="n">Reads</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">934456</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.0652</span>
<span class="n">Long</span><span class="o">-</span><span class="n">Range</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">2812664</span> <span class="o">/</span> <span class="mi">9050818</span> <span class="o">=</span> <span class="mf">0.3108</span>
<span class="n">Data</span> <span class="n">Usage</span> <span class="o">=</span> <span class="mi">9050818</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.6315</span>
</pre></div>
</div>
<p>The following table lists possible statistic names and their meanings:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 38%" />
<col style="width: 62%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Statistic Name</p></th>
<th class="head"><p>Meaning</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>000_SequencedReads</p></td>
<td><p>Total number of sequenced read pairs</p></td>
</tr>
<tr class="row-odd"><td><p>010_DoubleSideMappedReads</p></td>
<td><p>Number of read pairs of which both sides can be
uniquely mapped to the reference genome.</p></td>
</tr>
<tr class="row-even"><td><p>020_SingleSideMappedReads</p></td>
<td><p>Number of read pairs of which only one side can
be uniquely mapped to the reference genome.</p></td>
</tr>
<tr class="row-odd"><td><p>030_UnmappedReads</p></td>
<td><p>Number of read pairs of which neither side can
be uniquely mapped to the reference genome.</p></td>
</tr>
<tr class="row-even"><td><p>100_NormalPairs</p></td>
<td><p>Number of read pairs of which both sides can be
uniquely mapped.</p></td>
</tr>
<tr class="row-odd"><td><p>110_AfterFilteringReads</p></td>
<td><p>Number of read pairs that have passed all
filtering criteria.</p></td>
</tr>
<tr class="row-even"><td><p>120_SameFragmentReads</p></td>
<td><p>Number of read pairs of which both sides are
mapped to the same restriction fragment. Such
read pairs are filtered out in our pipeline.</p></td>
</tr>
<tr class="row-odd"><td><p>122_SelfLigationReads</p></td>
<td><p>Number of read pairs deriving from
self-circularized ligation product. The two sides
are mapped to the same restriction fragment and
face in opposite directions.</p></td>
</tr>
<tr class="row-even"><td><p>124_DanglingReads</p></td>
<td><p>Both sides of these read pairs are mapped to the
same fragment and face toward each other. There
can be many causes of such products, ranging from
low ligation efficiency to poor streptavidin
specificity.</p></td>
</tr>
<tr class="row-odd"><td><p>126_UnknownMechanism</p></td>
<td><p>Unknown sources of “120_SameFragmentReads”. Both
sides are mapped to the same strand.</p></td>
</tr>
<tr class="row-even"><td><p>310_DuplicatedRemoved</p></td>
<td><p>Number of read pairs from PCR products.</p></td>
</tr>
<tr class="row-odd"><td><p>400_TotalContacts</p></td>
<td><p>Number of read pairs from true contacts, i.e.,
the remaining read pairs after all filtering
processes.</p></td>
</tr>
<tr class="row-even"><td><p>410_IntraChromosomalReads</p></td>
<td><p>Number of intra-chromosomal contacts</p></td>
</tr>
<tr class="row-odd"><td><p>412_IntraLongRangeReads</p></td>
<td><p>Number of long-range contacts (genomic distance
&gt;= 20Kb)</p></td>
</tr>
<tr class="row-even"><td><p>412_IntraShortRangeReads</p></td>
<td><p>Number of short-range contacts (genomic distance
&lt; 20Kb)</p></td>
</tr>
<tr class="row-odd"><td><p>420_InterChromosomalReads</p></td>
<td><p>Number of inter-chromosomal contacts</p></td>
</tr>
</tbody>
</table>
<p>Note that we try to organize these statistics hierarchically using indentation,
so that, for example, “010_DoubleSideMappedReads”, “020_SingleSideMappedReads” and
“030_UnmappedReads” constitutes “000_SequencedReads”.</p>
<p>At the bottom of the statistic table, we include some important quality indicators:</p>
<ol class="arabic simple">
<li><p>Unique-Mapping Ratio. Low value of this metric indicates low sequencing quality,
sample contamination or incomplete genome assembly.</p></li>
<li><p>Self-Ligation Ratio.</p></li>
<li><p>Dangling-Reads Ratio.</p></li>
<li><p>Long-Range Ratio. Low value (&lt;0.15) of this metric indicates the failed experiment.</p></li>
</ol>
</section>
<section id="library-size-estimation">
<h2>Library-size Estimation<a class="headerlink" href="#library-size-estimation" title="Permalink to this headline">¶</a></h2>
<p>Dangling reads can be applied to estimate your library size in nature. Here’s an example
of size distribution of dangling read molecules for typical 300~500bp library:</p>
<img alt="_images/GM06990-HindIII-allReps-librarySize.png" class="align-center" src="_images/GM06990-HindIII-allReps-librarySize.png" />
<p>The inconsistency between this distribution and the experimental library size suggests
a failure in DNA size selection.</p>
</section>
<section id="read-pair-type-plotting">
<h2>Read-pair Type Plotting<a class="headerlink" href="#read-pair-type-plotting" title="Permalink to this headline">¶</a></h2>
<p>Read-pair type ratios will also be reported under <em>filtered-hg38</em>.
Intra-chromosomal contacts are broken down into four types: “left pair” (both sides map
to the reverse strand), “right pair” (both sides map to the forward strand), “inner pair”
(two sides map to different strands and point towards each other) and “outer pair” (two
sides map to different strands and point away from one another). If reads come from proximity
ligation, each pair type should account for roughly 25% of contacts. Thus, distance
at which the percentage of each type converges to 25% is a good indication of the minimum
distance at which it is meaningful to examine Hi-C contact patterns. Here’s an example
below:</p>
<img alt="_images/GM06990-HindIII-allReps-PairType.png" class="align-center" src="_images/GM06990-HindIII-allReps-PairType.png" />
<p>We can see a distinct turning point around 40Kb. While there may be several unknown mechanisms
making biases below this point, we should only consider contacts whose genomic distances
are greater than 40Kb in the following analysis.</p>
</section>
</section>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2021, XiaoTao Wang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.1.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/quality.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/XiaoTaoWang/HiC_pipeline" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>