
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Experiment Quality Assessment &#8212; runHiC 0.7.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">runHiC</a></h1>



<p class="blurb">A user-friendly Hi-C data processing software based on hiclib </p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=XiaoTaoWang&repo=HiC_pipeline&type=star&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>


<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="experiment-quality-assessment">
<h1>Experiment Quality Assessment<a class="headerlink" href="#experiment-quality-assessment" title="Permalink to this headline">¶</a></h1>
<p>In this tutorial, I will show you how runHiC can be used in data quality assessment.</p>
<p>All you need to type in is a sinlge line command after <em>runHiC filtering</em> or <em>runHiC pileup</em>
(refer to <a class="reference external" href="http://xiaotaowang.github.io/HiC_pipeline/quickstart.html">quickstart</a> for more details):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ runHiC quality -m datasets.tsv -L filtered-hg19
</pre></div>
</div>
<ul>
<li><p class="first"><code class="docutils literal"><span class="pre">-m/--metadata</span></code></p>
<p>The metadata data file name.</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">-L/--Locator</span></code></p>
<p>Path to the folder containing filtered HDF5 files.</p>
</li>
</ul>
<div class="section" id="statistic-table">
<h2>Statistic Table<a class="headerlink" href="#statistic-table" title="Permalink to this headline">¶</a></h2>
<p>In our example, statistic tables on sequencing reads for each SRA/FASTQ (level 1),
biological replicate (level 2) and cell type (level 3) will be generated under the
“filtered-hg19” folder.</p>
<p>Here’s a snapshot:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="mi">000</span><span class="n">_SequencedReads</span><span class="p">:</span>   <span class="mi">14332993</span>
        <span class="mi">010</span><span class="n">_UniqueMappedReads</span><span class="p">:</span>   <span class="mi">13053095</span>
        <span class="mi">020</span><span class="n">_LigationCounts</span><span class="p">:</span>   <span class="mi">3225524</span>
<span class="mi">100</span><span class="n">_DoubleUniqueMapped</span><span class="p">:</span>   <span class="mi">8260823</span>
        <span class="mi">110</span><span class="n">_AfterFilteringReads</span><span class="p">:</span>   <span class="mi">6836510</span>
        <span class="mi">120</span><span class="n">_SameFragmentReads</span><span class="p">:</span>   <span class="mi">1147363</span>
                <span class="mi">122</span><span class="n">_SelfLigationReads</span><span class="p">:</span>   <span class="mi">210202</span>
                <span class="mi">124</span><span class="n">_DanglingReads</span><span class="p">:</span>   <span class="mi">928888</span>
                <span class="mi">126</span><span class="n">_UnknownMechanism</span><span class="p">:</span>   <span class="mi">8273</span>
        <span class="mi">210</span><span class="n">_ExtraDanglingReads</span><span class="p">:</span>   <span class="mi">250193</span>
        <span class="mi">310</span><span class="n">_DuplicatedRemoved</span><span class="p">:</span>   <span class="mi">26757</span>
<span class="mi">400</span><span class="n">_TotalContacts</span><span class="p">:</span>   <span class="mi">6836510</span>
        <span class="mi">410</span><span class="n">_IntraChromosomalReads</span><span class="p">:</span>   <span class="mi">2753015</span>
                <span class="mi">412</span><span class="n">_IntraLongRangeReads</span><span class="p">(</span><span class="o">&gt;=</span><span class="mi">20</span><span class="n">Kb</span><span class="p">):</span>   <span class="mi">2402139</span>
                <span class="mi">412</span><span class="n">_IntraShortRangeReads</span><span class="p">(</span><span class="o">&lt;</span><span class="mi">20</span><span class="n">Kb</span><span class="p">):</span>   <span class="mi">350876</span>
        <span class="mi">420</span><span class="n">_InterChromosomalReads</span><span class="p">:</span>   <span class="mi">4083495</span>

<span class="n">Critical</span> <span class="n">Indicators</span><span class="p">:</span>
<span class="n">Double</span> <span class="n">Unique</span> <span class="n">Mapped</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">8260823</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.5764</span>
<span class="n">Ligation</span><span class="o">-</span><span class="n">Junction</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">3225524</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.2250</span>
<span class="n">Self</span><span class="o">-</span><span class="n">Ligation</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">210202</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.0147</span>
<span class="n">Dangling</span><span class="o">-</span><span class="n">Reads</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">928888</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.0648</span>
<span class="n">Long</span><span class="o">-</span><span class="n">Range</span> <span class="n">Ratio</span> <span class="o">=</span> <span class="mi">2402139</span> <span class="o">/</span> <span class="mi">6836510</span> <span class="o">=</span> <span class="mf">0.3514</span>
<span class="n">Data</span> <span class="n">Usage</span> <span class="o">=</span> <span class="mi">6836510</span> <span class="o">/</span> <span class="mi">14332993</span> <span class="o">=</span> <span class="mf">0.4770</span>
</pre></div>
</div>
<p>The following table lists possible statistic names and their meanings:</p>
<p>Note that we try to organize these statistics hierarchically using indentation,
so “010_UniqueMappedReads” and “020_LigationCounts” are parts of “000_SequencedReads”,
similarly, “122_SelfLigationReads”, “124_DanglingReads” and “126_UnknownMechanism”
constitute “120_SameFragmentReads”.</p>
<p>At the bottom of the statistic table, we include some important quality indicators:</p>
<ol class="arabic simple">
<li>Unique-Mapping Ratio. Low value of this metric indicates low sequencing quality,
sample contamination or incomplete genome assembly.</li>
<li>Ligation-Junction Ratio. Low value of this metric suggests the failure of ligation.</li>
<li>Self-Ligation Ratio.</li>
<li>Dangling-Reads Ratio.</li>
<li>Long-Range Ratio. Low value (&lt;0.15) of this metric indicates the failed experiment.</li>
</ol>
</div>
<div class="section" id="library-size-estimation">
<h2>Library-size Estimation<a class="headerlink" href="#library-size-estimation" title="Permalink to this headline">¶</a></h2>
<p>Dangling reads can be applied to estimate your library size in nature. Here’s an example
of size distribution of dangling read molecules for typical 300~500bp library:</p>
<img alt="./_static/GM06990-HindIII-allReps-librarySize.png" class="align-center" src="./_static/GM06990-HindIII-allReps-librarySize.png" />
<p>The inconsistency between this distribution and the experimental library size suggests
a failure in DNA size selection.</p>
</div>
<div class="section" id="ligation-efficiency">
<h2>Ligation Efficiency<a class="headerlink" href="#ligation-efficiency" title="Permalink to this headline">¶</a></h2>
<p>Excessive dangling reads may result from low ligation efficiency or poor streptavidin
specificity. Can we further discriminate these two mechanisms? Intuitively, if one of
reads of most read pairs locate near a restriction site, the former (low ligation efficiency)
is more likely to be the cause, so we also plot the distribution of the relative start
sites for dangling reads:</p>
<img alt="./_static/GM06990-HindIII-allReps-danglingStart.png" class="align-center" src="./_static/GM06990-HindIII-allReps-danglingStart.png" />
<p>Here, the majority of these read pairs have one of their read starting near a restriction
site, therefore, ligation efficiency could be a good explain.</p>
</div>
<div class="section" id="read-pair-type-plotting">
<h2>Read-pair Type Plotting<a class="headerlink" href="#read-pair-type-plotting" title="Permalink to this headline">¶</a></h2>
<p>Read-pair type ratios will be reported in line-plot manner for each biological
replicate (level 1) and cell type (level 2) under filtered-hg19 too. Intra-chromosomal
contacts are broken down into four types: “left pair” (both reads map to the reverse
strand), “right pair” (both reads map to the forward strand), “inner pair” (reads map
to different strands and point towards each other) and “outer pair” (reads map to
different strands and point away from one another). If reads come from proximity
ligation, each pair type should account for roughly 25% of contacts. Thus, distance
at which the percentage of each type converges to 25% is a good indication of the minimum
distance at which it is meaningful to examine Hi-C contact patterns. Here’s an example
below:</p>
<img alt="./_static/GM06990-HindIII-allReps-PairType.png" class="align-center" src="./_static/GM06990-HindIII-allReps-PairType.png" />
<p>We can see a distinct turning point around 20Kb. While there may be several unknown mechanisms
making biases below this point, we should only consider contacts whose genomic distances
are greater than 20Kb in the following analysis.</p>
</div>
</div>
<div class="section" id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h1>
<table class="docutils footnote" frame="void" id="id1" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td>Rao SS, Huntley MH, Durand NC et al. A 3D Map of the Human Genome at Kilobase Resolution
Reveals Principles of Chromatin Looping. Cell, 2014, 159(7):1665-80.</td></tr>
</tbody>
</table>
</div>


          </div>
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, XiaoTao Wang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/quality.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/XiaoTaoWang/HiC_pipeline" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>