
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Quick Start &#8212; runHiC 0.8.4 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Parallel Computation" href="parallel.html" />
    <link rel="prev" title="Install runHiC" href="install.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">runHiC</a></h1>



<p class="blurb">A user-friendly Hi-C data processing software supporting distributed computation</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=XiaoTaoWang&repo=HiC_pipeline&type=star&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Install runHiC</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick Start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-preparation">Data Preparation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mapping">Mapping</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#create-the-meta-data-file">Create the Meta Data File</a></li>
<li class="toctree-l3"><a class="reference internal" href="#runhic-command">runHiC Command</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#filtering">Filtering</a></li>
<li class="toctree-l2"><a class="reference internal" href="#binning">Binning</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pileup">Pileup</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="parallel.html">Parallel Computation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality.html">Experiment Quality Assessment</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Release Notes</a></li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="quick-start">
<h1>Quick Start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h1>
<p>This tutorial covers the basic use of 4 subcommands (<em>mapping</em>, <em>filtering</em>,
<em>binning</em>, and <em>pileup</em>) provided by runHiC. We will first download an example
Hi-C data set and corresponding reference genome data. Then we will process the
Hi-C data step by step from raw sequencing reads (.sra, .fastq, and .fastq.gz)
to the ICE-corrected contact matrices. Lastly, we will demonstrate how to streamline
the processing pipeline by using the <em>pileup</em> subcommand.</p>
<div class="section" id="data-preparation">
<h2>Data Preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">¶</a></h2>
<p>First of all, let’s make a temporary blank folder somewhere on your system, and change your
working directory to it:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir data
$ mkdir workspace
$ ls -lh

total 0
drwxr-xr-x  2 xtwang  staff    64B Sep 16 11:24 data
drwxr-xr-x  2 xtwang  staff    64B Sep 16 11:25 workspace
</pre></div>
</div>
<p>During this tutorial, all input data including Hi-C raw sequencing data and
the reference genome data will be placed under the <em>data</em> sub-folder, and
<em>runHiC</em> will be run under the <em>workspace</em> sub-folder.</p>
<p>Download the example Hi-C data set using the <em>prefetch</em> command of the SRA toolkit:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd data
$ mkdir HiC-SRA
$ cd HiC-SRA
$ prefetch SRR027956
$ prefetch SRR027958
$ ls -lh

total 1.4G
-rw-r--r-- 1 xtwang staff 623M Sep 16 11:49 SRR027956.sra
-rw-r--r-- 1 xtwang staff 783M Sep 16 11:53 SRR027958.sra
</pre></div>
</div>
<p>To demonstrate all read formats supported by <em>runHiC</em>, let’s first dump reads
using the <em>fastq-dump</em> command, and then compress the FASTQ files using <em>gzip</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ for i in ./*.sra; do fastq-dump --split-files $i; done
$ for i in ./*.fastq; do gzip -c $i &gt; `basename $i`.gz; done
</pre></div>
</div>
<p><em>runHiC</em> currently supports three read formats: FASTQ, compressed FASTQ (.fastq.gz),
and SRA(Sequence Read Archive):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ..
$ mkdir HiC-FASTQ
$ mkdir HiC-gzip
$ mv ./HiC-SRA/*.fastq ./HiC-FASTQ
$ mv ./HiC-SRA/*.gz ./HiC-gzip
</pre></div>
</div>
<p>Download the reference genome (hg38) data from UCSC:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir hg38
$ cd hg38
$ wget ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/*
</pre></div>
</div>
<p>Above commands can be modified to download any other genomes available in UCSC
by replacing “hg38” with the desired reference genome release name.</p>
<p>We will only include sequences that have been completely assembled. Open
a python interpreter and follow the commands below:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">glob</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">)))</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span><span class="s1">&#39;M&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pool</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chr</span><span class="si">{0}</span><span class="s1">.fa.gz&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*.fa.gz&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="ow">not</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">pool</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exit</span><span class="p">()</span>
</pre></div>
</div>
<p>Finally uncompress the .gz files and merge all chromosomes into hg38.fa:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ gunzip *.gz
$ cat *.fa &gt; hg38.fa
$ cd ../..
</pre></div>
</div>
</div>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<p>The first processing stage of <em>runHiC</em> is conducted by the <em>mapping</em> subcommand,
which maps raw sequencing reads to the reference genome and parses the resulted
SAM/BAM alignments into <a class="reference external" href="https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md">.pairs</a>.</p>
<div class="section" id="usage">
<h3>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">runHiC</span> <span class="pre">mapping</span> <span class="pre">[options]</span></code></p>
</div>
<div class="section" id="create-the-meta-data-file">
<h3>Create the Meta Data File<a class="headerlink" href="#create-the-meta-data-file" title="Permalink to this headline">¶</a></h3>
<p>Another thing you need to do is to prepare a meta data file describing your Hi-C
data under the <em>workspace</em> sub-folder:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd workspace
$ vim datasets.tsv
</pre></div>
</div>
<p>Create a TXT file called “datasets.tsv” by <code class="docutils literal notranslate"><span class="pre">vim</span></code> and fill in the following content:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR027956</span> <span class="n">GM06990</span> <span class="n">R1</span> <span class="n">HindIII</span>
<span class="n">SRR027958</span> <span class="n">GM06990</span> <span class="n">R2</span> <span class="n">HindIII</span>
</pre></div>
</div>
<p>The meta data file should contain 4 columns: prefix of the SRA file name (in the
case of the FASTQ read format, it should be the leading part of the file name
apart from the “_1.fastq” or “_2.fastq” substring), cell line name, biological
replicate label, and the restriction enzyme name.</p>
</div>
<div class="section" id="runhic-command">
<h3>runHiC Command<a class="headerlink" href="#runhic-command" title="Permalink to this headline">¶</a></h3>
<p>Now type and execute the command below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC mapping -p ../data/ -g hg38 -f HiC-SRA -F SRA -A bwa-mem -t 10 -O BAM --include-readid --include-sam --drop-seq --chunkSize 1500000 --logFile runHiC-mapping.log
</pre></div>
</div>
<p>For FASTQ and the compressed FASTQ format, replace “HiC-SRA” with “HiC-FASTQ”
or “HiC-gzip”, and reset “-F” argument accordingly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC mapping -p ../data/ -g hg38 -f HiC-gzip -F FASTQ -A bwa-mem -t 10 -O BAM --include-readid --include-sam --drop-seq --chunkSize 1500000 --logFile runHiC-mapping.log
</pre></div>
</div>
<p>Two sub-folders named <em>alignments-hg38</em> and <em>pairs-hg38</em> will be created under current
working directory (<em>workspace</em>):</p>
<ol class="arabic simple">
<li><p>Read pairs will be mapped to the <em>hg38</em> reference genome with <code class="docutils literal notranslate"><span class="pre">bwa</span> <span class="pre">mem</span></code>, and the
alignment results will be reported in BAM format and placed under <em>alignments-hg38</em>.</p></li>
<li><p>BAM files will be parsed into <a class="reference external" href="https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md">.pairs</a>
using <a class="reference external" href="https://github.com/mirnylab/pairtools">pairtools</a> under <em>pairs-hg38</em>.</p></li>
</ol>
<p><em>runHiC</em> now supports two read aligners, <em>bwa</em> and <em>minimap2</em>. You can switch it by the <code class="docutils literal notranslate"><span class="pre">-A/--aligner</span></code>
argument. Generally, <em>bwa</em> performs better on short reads (&lt;=100bp), and <em>minimap2</em> runs
faster and is similiarly accurate as <em>bwa</em> on longer reads (&gt;100bp).</p>
<p>For alignment format, you can choose between SAM(TXT) and BAM(Binary) with the <code class="docutils literal notranslate"><span class="pre">-O/--outformat</span></code>
argument.</p>
<p>During the alignment parsing, <em>runHiC</em> detects ligation junctions, marks various situations
(Unmapped, Multimapped, Multiple ligations-Walks, and valid Single ligations), and sort
pairs for further analysis. In this example, .pairsam.gz files under <em>pairs-hg38</em> are
valid .pairs files proposed by the <a class="reference external" href="https://www.4dnucleome.org">4DN</a> consortium. By default,
it will only contain 7 columns: chr1, pos1, chr2, pos2, strand1, strand2, and pair_type;
if you add <code class="docutils literal notranslate"><span class="pre">--include-readid</span></code> on the command, you will get an additional “readID” column;
if you specify <code class="docutils literal notranslate"><span class="pre">--include-sam</span></code>, two extra columns “sam1” and “sam2” will be added to store
the original alignments; if you add <code class="docutils literal notranslate"><span class="pre">--drop-seq</span></code>, SEQ and QUAL will be removed from the sam
fields to save the disk space.</p>
<p><em>runHiC</em> uses a rotating file for logging. According to our settings, when the size of
“runHiC.log” gets about 100K, it is closed and renamed to “runHiC.log.1”. At the same
time, a new file “runHiC.log” is silently opened for output. In a word, the system saves
old log files by appending the extensions “.1”, “.2” etc., and the current log is always
written to “runHiC.log”.</p>
</div>
</div>
<div class="section" id="filtering">
<h2>Filtering<a class="headerlink" href="#filtering" title="Permalink to this headline">¶</a></h2>
<p>The <em>filtering</em> subcommand of <em>runHiC</em> is designed to perform basic filtering procedures on
the aligned read pairs:</p>
<ol class="arabic simple">
<li><p>Remove redundant PCR artifacts.</p></li>
<li><p>Remove the read pair that maps to the same restriction fragment.</p></li>
</ol>
<p>During the filtering process, <em>runHiC</em> also records read-level, fragment-level and the
contact-level statistics for quality assessment of your Hi-C data.
(See <a class="reference external" href="http://xiaotaowang.github.io/HiC_pipeline/quality.html">quality</a>)</p>
<p>Here’s the command you should type in the terminal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC filtering --pairFolder pairs-hg38/ --logFile runHiC-filtering.log --nproc 10
</pre></div>
</div>
<p>That will create a new sub-folder named <em>filtered-hg38</em>. Please find the final valid
contact pairs in <a href="#id2"><span class="problematic" id="id3">*</span></a>.pairs.gz files. If you specified <code class="docutils literal notranslate"><span class="pre">--include-sam</span></code> when you ran
<code class="docutils literal notranslate"><span class="pre">runHiC</span> <span class="pre">mapping</span></code>, it will also output a .bam file accompanying each .pairs.gz file
to store alignments that passed all filtering criteria.</p>
</div>
<div class="section" id="binning">
<h2>Binning<a class="headerlink" href="#binning" title="Permalink to this headline">¶</a></h2>
<p>At this stage, an .mcool file will be produced under the <em>coolers-hg38</em> sub-folder for each
.pairs.gz file using <a class="reference external" href="https://cooler.readthedocs.io/en/latest/">cooler</a>. The mcool format
is the official Hi-C data format for the <a class="reference external" href="https://data.4dnucleome.org/resources/data-analysis/hi_c-processing-pipeline">4DN consortium</a>
and can be visualized using <a class="reference external" href="https://docs.higlass.io/">HiGlass</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC binning -f filtered-hg38/ --logFile runHiC-binning.log --nproc 10
</pre></div>
</div>
</div>
<div class="section" id="pileup">
<h2>Pileup<a class="headerlink" href="#pileup" title="Permalink to this headline">¶</a></h2>
<p><em>runHiC</em> also provides a handy subcommand called “pileup” by which you can perform all
processing steps above using the single-line command below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC pileup -p ../data/ -g hg38 -f HiC-SRA -F SRA -A bwa-mem -t 10 -O BAM --include-readid --include-sam --drop-seq --chunkSize 1500000 --logFile runHiC.log
</pre></div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2021, XiaoTao Wang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/quickstart.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/XiaoTaoWang/HiC_pipeline" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>