
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Quick Start &#8212; runHiC 0.7.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Parallel Computation" href="parallel.html" />
    <link rel="prev" title="Install runHiC" href="install.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">runHiC</a></h1>



<p class="blurb">A user-friendly Hi-C data processing software based on hiclib </p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=XiaoTaoWang&repo=HiC_pipeline&type=star&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Install runHiC</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick Start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-preparation">Data Preparation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mapping">Mapping</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bowtie2-path">Bowtie2 Path</a></li>
<li class="toctree-l3"><a class="reference internal" href="#write-the-meta-data-file">Write the Meta Data File</a></li>
<li class="toctree-l3"><a class="reference internal" href="#runhic-command">runHiC Command</a></li>
<li class="toctree-l3"><a class="reference internal" href="#access-data-from-hdf5">Access Data from HDF5</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#filtering">Filtering</a></li>
<li class="toctree-l2"><a class="reference internal" href="#binning">Binning</a></li>
<li class="toctree-l2"><a class="reference internal" href="#correcting">Correcting</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tosparse">tosparse</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pileup">Pileup</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reference">Reference</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="parallel.html">Parallel Computation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality.html">Experiment Quality Assessment</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Release Notes</a></li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="quick-start">
<h1>Quick Start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h1>
<p>This tutorial covers the basic use of 6 subcommands (<em>mapping</em>, <em>filtering</em>,
<em>binning</em>, <em>correcting</em>, <em>tosparse</em>, and <em>pileup</em>) provided by runHiC. We
will first download an example Hi-C data set and corresponding reference
genome data. Then we will use runHiC to process the Hi-C data step by step
from raw sequencing reads (.sra, .fastq, .fastq.gz) to the corrected
contact matrices. Lastly, we will demonstrate how to streamline the processing
pipeline by <em>pileup</em> subcommand.</p>
<div class="section" id="data-preparation">
<h2>Data Preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">¶</a></h2>
<p>Suppose you are still in the <em>runHiC</em> distribution root folder, change your
current working directory to the sub-folder <em>demo</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd demo
$ mkdir data
$ ls -lh

total 16K
drwxr-x--- 2 xtwang CPeng 4.0K Oct 29 10:03 data
drwxr-x--- 2 xtwang CPeng 4.0K Oct 29 10:01 workspace
</pre></div>
</div>
<p>During this tutorial, all input data including Hi-C raw sequencing data and
the reference genome data will be placed under the <em>data</em> sub-folder, and
<em>runHiC</em> will be run under the <em>workspace</em> sub-folder.</p>
<p>Download the example Hi-C data set from a human cell line GM06990:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd data
$ mkdir HiC-SRA
$ cd HiC-SRA
$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX011/SRX011608/SRR027956/SRR027956.sra -O SRR027956.sra
$ wget ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX/SRX011/SRX011610/SRR027958/SRR027958.sra -O SRR027958.sra
$ ls -lh

total 1.4G
-rw-r----- 1 xtwang CPeng 623M Oct 29 10:07 SRR027956.sra
-rw-r----- 1 xtwang CPeng 783M Oct 29 10:10 SRR027958.sra
</pre></div>
</div>
<p>Reads in SRR027956.sra and SRR027958.sra are sequenced from two biological replicates,
respectively.</p>
<p>To demonstrate all read formats supported by <em>runHiC</em>, let’s first dump reads
with <em>fastq-dump</em> and then compress the FASTQ files with gzip:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ for i in ./*.sra; do fastq-dump --split-3 $i; done
$ for i in ./*.fastq; do gzip -c $i &gt; `basename $i`.gz; done
</pre></div>
</div>
<p><em>runHiC</em> currently supports 3 sequencing read format: FASTQ, compressed FASTQ with gzip,
and SRA(Sequence Read Archive):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ..
$ mkdir HiC-FASTQ
$ mkdir HiC-gzip
$ mv ./HiC-SRA/*.fastq ./HiC-FASTQ
$ mv ./HiC-SRA/*.gz ./HiC-gzip
</pre></div>
</div>
<p>Download the reference genome (hg19) data from UCSC:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir hg19
$ cd hg19
$ wget ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/*
$ wget ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/gap.txt.gz
</pre></div>
</div>
<p>Above commands can be modified to download the data for any other genomes
available in UCSC by replacing “hg19” with the desired reference genome
release name.</p>
<p>We will only include sequences that have been completely assembled. Open
a python interpreter and follow the commands below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">glob</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">labels</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span><span class="s1">&#39;M&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pool</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chr</span><span class="si">{0}</span><span class="s1">.fa.gz&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*.fa.gz&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="ow">not</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">pool</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exit</span><span class="p">()</span>
</pre></div>
</div>
<p>Finally, uncompress the .gz files to finish this section:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ gunzip *.gz
$ cd ../..
</pre></div>
</div>
</div>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<p>The first processing stage of <em>runHiC</em> is conducted by the <em>mapping</em> subcommand,
which maps raw sequencing reads to the reference genome and assigns aligned
reads to the restriction fragments.</p>
<p><em>runHiC</em> also records the read-level statistics at this stage for quality
assessment of your Hi-C data. (See <a class="reference external" href="http://xiaotaowang.github.io/HiC_pipeline/quality.html">quality</a>)</p>
<div class="section" id="usage">
<h3>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">runHiC</span> <span class="pre">mapping</span> <span class="pre">[options]</span></code></p>
</div>
<div class="section" id="bowtie2-path">
<h3>Bowtie2 Path<a class="headerlink" href="#bowtie2-path" title="Permalink to this headline">¶</a></h3>
<p><em>runHiC</em> calls the <em>bowtie2</em> read alignment software during the <em>mapping</em> stage,
so you need to tell <em>runHiC</em> where <em>bowtie2</em> is installed in your system:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ which bowtie2

~/Tools/anaconda2/bin/bowtie2
</pre></div>
</div>
</div>
<div class="section" id="write-the-meta-data-file">
<h3>Write the Meta Data File<a class="headerlink" href="#write-the-meta-data-file" title="Permalink to this headline">¶</a></h3>
<p>Another thing you need to do is to prepare a meta data file describing your Hi-C
data under the <em>workspace</em> sub-folder:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd workspace
$ cat datasets.tsv
</pre></div>
</div>
<p>The meta data file should contain 4 columns: prefix of the SRA file name (in the
case of the FASTQ read format, it should be the leading part of the file name
apart from the “_1.fastq” or “_2.fastq” substring), cell line name, biological
replicate label, and the restriction enzyme name:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SRR027956</span> <span class="n">GM06990</span> <span class="n">R1</span> <span class="n">HindIII</span>
<span class="n">SRR027958</span> <span class="n">GM06990</span> <span class="n">R2</span> <span class="n">HindIII</span>
</pre></div>
</div>
</div>
<div class="section" id="runhic-command">
<h3>runHiC Command<a class="headerlink" href="#runhic-command" title="Permalink to this headline">¶</a></h3>
<p>Now type in the command below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC mapping -m datasets.tsv -p ../data -g hg19 -G gap.txt -f HiC-SRA -F SRA -b ~/Tools/anaconda2/bin/bowtie2 -t 20 --removeInters --logFile runHiC-mapping.log
</pre></div>
</div>
<p>For FASTQ and the compressed FASTQ format, just replace “HiC-SRA” with “HiC-FASTQ” or “HiC-gzip”,
and reset “-F” argument correspondingly:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC mapping -m datasets.tsv -p ../data -g hg19 -G gap.txt -f HiC-gzip -F FASTQ -b ~/Tools/anaconda2/bin/bowtie2 -t 20 --removeInters --logFile runHiC-mapping.log
</pre></div>
</div>
<ul>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-m/--metadata</span></code></p>
<p>The metadata data file name.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-p/--dataFolder</span></code></p>
<p>Path to the root folder containing both Hi-C sequencing data and the reference
genome data.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-g/--genomeName</span></code></p>
<p>Name of the reference genome. (Or name of the folder containing the reference
genome data)</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-G/--gapFile</span></code></p>
<p>Name of the decompressed gap file downloaded from UCSC. If runHiC fails to find
it, a dummy one will be generated in the specified genome folder (see <code class="docutils literal notranslate"><span class="pre">-g/--genomeName</span></code>).</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-f/--fastqDir</span></code></p>
<p>Name of the folder containing the Hi-C raw sequencing data.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-F/Format</span></code></p>
<p>Format of the sequencing data. SRA or FASTQ.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-b/--bowtiePath</span></code></p>
<p>Path to the bowtie2 executable, see <a class="reference external" href="http://xiaotaowang.github.io/HiC_pipeline/quickstart.html#bowtie2-path">Bowtie2 Path</a>.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-t/--threads</span></code></p>
<p>Number of the bowtie2 threads.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">--removeInters</span></code></p>
<p>Remove intermediate results if specified.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">--logFile</span></code></p>
<p>Log file name.</p>
</li>
</ul>
<p>During the execution of <code class="docutils literal notranslate"><span class="pre">runHiC</span> <span class="pre">mapping</span></code>, two new sub-folders named <em>bams-hg19</em> and
<em>hdf5-hg19</em> are created under current working directory (<em>workspace</em>). The read pairs
are mapped to the <em>hg19</em> reference genome in an iterative way with <em>bowtie2</em>. <a class="footnote-reference" href="#id9" id="id2">[1]</a>
The alignment results are stored in the BAM format and placed under <em>bams-hg19</em>. Then
BAM files of corresponding read pairs are parsed together and outputed into HDF5 files
(suffixed with .hdf5) under <em>hdf5-hg19</em>.</p>
<p>runHiC uses a rotating file for logging. According to our settings, when the size of
“runHiC.log” gets about 100K, it is closed and renamed to “runHiC.log.1”. At the same
time, a new file “runHiC.log” is silently opened for output. In a word, the system saves
old log files by appending the extensions “.1”, “.2” etc., and the current log is always
written to “runHiC.log”.</p>
</div>
<div class="section" id="access-data-from-hdf5">
<h3>Access Data from HDF5<a class="headerlink" href="#access-data-from-hdf5" title="Permalink to this headline">¶</a></h3>
<p>You can extract data from HDF5 files via <em>mirnylib</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mirnylib</span> <span class="k">import</span> <span class="n">h5dict</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lib</span> <span class="o">=</span> <span class="n">h5dict</span><span class="o">.</span><span class="n">h5dict</span><span class="p">(</span><span class="s1">&#39;./hdf5-hg19/SRR027956/SRR027956.hdf5&#39;</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lib</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[u&#39;chrms1&#39;, u&#39;chrms2&#39;, u&#39;cuts1&#39;, u&#39;cuts2&#39;, u&#39;downrsites1&#39;, u&#39;downrsites2&#39;, u&#39;misc&#39;, u&#39;rfragIdxs1&#39;, u&#39;rfragIdxs2&#39;, u&#39;rsites1&#39;, u&#39;rsites2&#39;, u&#39;strands1&#39;, u&#39;strands2&#39;, u&#39;uprsites1&#39;, u&#39;uprsites2&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="filtering">
<h2>Filtering<a class="headerlink" href="#filtering" title="Permalink to this headline">¶</a></h2>
<p>The <em>filtering</em> subcommand of <em>runHiC</em> is designed to perform some basic filtering on
the aligned read pairs: <a class="footnote-reference" href="#id9" id="id3">[1]</a></p>
<ol class="arabic simple">
<li>Remove the read pair that maps to the same restriction fragment.</li>
<li>Remove redundant PCR artifacts.</li>
</ol>
<p>During the filtering process, <em>runHiC</em> also records the fragment-level and the
contact-level statistics for quality assessment of your Hi-C data. (See <a class="reference external" href="http://xiaotaowang.github.io/HiC_pipeline/quality.html">quality</a>)</p>
<p>Data from the same biological replicate (or optionally all replicates of the same cell
line) are merged in this processing stage.</p>
<p>Here’s the command you should type in the terminal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC filtering -m datasets.tsv --HDF5 hdf5-hg19 --libSize 500 --duplicates -l 2
</pre></div>
</div>
<ul>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-m/--metadata</span></code></p>
<p>The metadata data file name.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">--HDF5</span></code></p>
<p>Path to the root folder of HDF5 files generated in the <em>mapping</em> stage.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">--libSize</span></code></p>
<p>Maximum length of molecules in your Hi-C library.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">--duplicates</span></code></p>
<p>Remove redundant PCR artifacts if specified.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-l/--level</span></code></p>
<p>Merging level. 1: Merge data from the same biological replicate; 2: Merge data from
all replicates of the same cell line.</p>
</li>
</ul>
<p>That will create a new sub-folder named <em>filtered-hg19</em> in which the filtered data
(stored in the HDF5 format, suffixed with .hdf5) reside. (See <a class="reference external" href="http://xiaotaowang.github.io/HiC_pipeline/quickstart.html#access-data-from-hdf5">Access Data from HDF5</a>)</p>
</div>
<div class="section" id="binning">
<h2>Binning<a class="headerlink" href="#binning" title="Permalink to this headline">¶</a></h2>
<p>This processing stage is conducted by the <em>binning</em> subcommand.</p>
<p>By partitioning the linear genome into fixed-size bins (intervals), the number of contacts
(the filtered read pairs) can be counted between any two bins (reads are assigned to bins
by the centers of their fragments), which results in a “contact” matrix.</p>
<p><em>runHiC</em> supports two modes of binning: <strong>wholeGenome</strong> builds a single genome-versus-genome
contact matrix, and <strong>byChromosome</strong> builds chromosome-versus-chromosome contact matrices.</p>
<p>Type in the command below to carry on our analyzing on the example data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC binning -f filtered-hg19 -M byChromosome -R 2000000
</pre></div>
</div>
<ul>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-f/--filteredDir</span></code></p>
<p>Path to the hdf5 files generated during the <em>filtering</em> stage. Wild cards are allowed.
If a path points to a folder, the <em>binning</em> procedure will be performed on each hdf5
file under that folder.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-M/--mode</span></code></p>
<p>Mode label for building contact matrices. <em>wholeGenome</em> or <em>byChromosome</em>.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-R/--resolution</span></code></p>
<p>Resolution of the generated contact matrices. Unit: bp</p>
</li>
</ul>
<p>After this command, a new sub-folder named <em>Raw-hg19</em> will be created under current
working directory. The original contact matrices are also stored in hdf5 files suffixed
with “.hm”. (See <a class="reference external" href="http://xiaotaowang.github.io/HiC_pipeline/quickstart.html#access-data-from-hdf5">Access Data from HDF5</a>)</p>
</div>
<div class="section" id="correcting">
<h2>Correcting<a class="headerlink" href="#correcting" title="Permalink to this headline">¶</a></h2>
<p>Hi-C data can contain many different biases, some from known origin and others from an
unknown origin. There are two general approaches to Hi-C bias correction: explicit-factor
methods <a class="footnote-reference" href="#id10" id="id7">[2]</a> and matrix balancing methods <a class="footnote-reference" href="#id9" id="id8">[1]</a>. <em>runHiC</em> (hiclib) uses the matrix balancing
algorithm called Sinkhorn–Knopp for bias correction.</p>
<p>To correct our sample Hi-C data, type in the command below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC correcting -H Raw-hg19
</pre></div>
</div>
<ul>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-H/--Heatmap</span></code></p>
<p>Path to the original contact matrices generated during the last stage. Wild cards are
allowed. If a path points to a folder, bias correction will be performed on each “.hm”
file under that folder.</p>
</li>
</ul>
<p>After that, a new sub-folder named <em>Corrected-hg19</em> with corrected contact matrices will
be created under current working directory.</p>
</div>
<div class="section" id="tosparse">
<h2>tosparse<a class="headerlink" href="#tosparse" title="Permalink to this headline">¶</a></h2>
<p>Contact matrices generated by <em>binning</em> and <em>correcting</em> subcommand are represented as
dense matrices, which are memory-consuming for large genomes (such as human and mouse)
and high resolutions (40 Kb) when used in other calculations (such as TAD and loop identifications).
This subcommand converts intra-chromosomal contact matrices into sparse ones.</p>
<p>To convert the dense matrices generated just now into well-known CSR (Compressed Sparse
Row) matrices, type in the command below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC tosparse -H Corrected-hg19 --csr
</pre></div>
</div>
<p>– <code class="docutils literal notranslate"><span class="pre">-H/--cHeatMap</span></code></p>
<blockquote>
<div>Path to the dense matrix files from <em>binning</em> or <em>correcting</em>. Wild cards are allowed.
If a folder name is provided, conversion will be performed on each “.hm” file under
that folder.</div></blockquote>
<p>– <code class="docutils literal notranslate"><span class="pre">--csr</span></code></p>
<blockquote>
<div>If specified, dense matrices are converted into <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/sparse.html">CSR (Compressed Sparse Row)</a> matrices,
a <a class="reference external" href="http://xiaotaowang.github.io/TADLib/hitad.html#transform-txt-into-npz">customized numpy structured array</a> is applied otherwise.</div></blockquote>
<p>Sparse matrices are organized in <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.savez.html">npz</a> format
using chromosome labels as the keys. Here, you can find them under the “Corrected-hg19” folder.</p>
</div>
<div class="section" id="pileup">
<h2>Pileup<a class="headerlink" href="#pileup" title="Permalink to this headline">¶</a></h2>
<p><em>runHiC</em> also provides a handy subcommand called “pileup” by which you can perform all
processing steps above with single-line command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ runHiC pileup -p ../data -g hg19 -f HiC-SRA -F SRA -b ~/Tools/anaconda2/bin/bowtie2 -t 20 --removeInters -M byChromosome -R 2000000
</pre></div>
</div>
</div>
<div class="section" id="reference">
<h2>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h2>
<table class="docutils footnote" frame="void" id="id9" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td><em>(<a class="fn-backref" href="#id2">1</a>, <a class="fn-backref" href="#id3">2</a>, <a class="fn-backref" href="#id8">3</a>)</em> Imakaev M, Fudenberg G, McCord RP et al. Iterative correction of Hi-C data
reveals hallmarks ofchromosome organization. Nat Methods, 2012, 9(10): 999-1003.</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id10" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id7">[2]</a></td><td>Yaffe E, Tanay A. Probabilistic modeling of Hi-C contact maps eliminates
systematic biases to characterize global chromosomal architecture. Nat Genet,
2011, 43(11): 1059-65.</td></tr>
</tbody>
</table>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, XiaoTao Wang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.11</a>
      
      |
      <a href="_sources/quickstart.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/XiaoTaoWang/HiC_pipeline" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>